proposals/dgl-operator-proposal.md

```diff
@@ -0,0 +1,248 @@
+    2. `ipconfig`: A text file lists the worker pods' name, ip, port (used in training phase of dgl server) and the available slots (GPUs) in each pod
+2. Create the RBAC resources (`Role`, `ServiceAccount`, `RoleBinding`) to allow remote execution (`pods/exec`).
+3. Create worker `Pods` that reaches the desired replicas. The command is set to sleep forever.
+4. Create launcher `Pods`. And `initContainer` (`kubectl-delivery`) in `launcher` pod takes the responsibility to:
+    1. wait for all worker pods are ready
+    2. `kubectl` binary is delivered to an `emptyDir` volume
+5. Take off main container of the launcher `Pod`, and run the `dglrun` command and arguments.
+    1. `dglrun`: A must-have binary file that can take the control of DGL distributed training.
+    2. The launcher is very light weight.
+6. When `DGLJob` finishes, it will clean all workers.
+
+![DGL Operator-wise Diagram](diagrams/dgl-operator-1.png)
+
+### Training-wise
+
+**Default communication backend:** gloo
+
+**How does DGL implement distributed training**
+
+- data parallelism: partitioned sub-graph is like the divided data
+- the allreduce way: updating dense parameters is used DDP in PyTorch
+
+**How do we leverage DGL to run distributed training on Kubernetes**
+
+Prerequisite: the dataset must be included in worker `Pod` image, like [OGB](https://github.com/snap-stanford/ogb).
+
+0. When `kubectl-delivery` container is finished, the main container of launcher `Pod` starts running.
+1. In the main container, `dglrun` invokes a partition commond to do distributed partitioning on every worker `Pod`. Dive in [this approach](https://docs.dgl.ai/guide/distributed-preprocessing.html#distributed-partitioning).
+2. After finished distributed partitioning, `dglrun` invokes distributed training on every worker `Pod`. Guide [here](https://github.com/dmlc/dgl/tree/master/examples/pytorch/graphsage/experimental#step-3-launch-distributed-jobs).
+
+![DGL Training-wise Diagram](diagrams/dgl-operator-2.png)
+
+## Alternatives Considered
+
+### DGL Operator vs MPI Operator
+
+Currently, for distributed partitioning, DGL uses `ParMETIS` to split big graphs up and store in parallel on each worker `Pods`, which is empowered by MPI. For distributed training, DGL uses PyTorch's `DDP` in dense parameters updating, and self-developed methods in other scenarios.
+
+So in this manner, a fully-informed user, who understand any details about DGL and Operator philosophy, can hack all DGL workloads (distributed partitioning & training) on MPI Operator. However, this hack approach is not a good idea, because 1) user need to revise `ipconfig` processing in DGL library to adapt MPI Operator's `hostfile`; 2) user may need involve intrusive code in partition and train script.
+
+Therefore, DGL Operator will follow the community, to provide native DGL distributed training experience. There is no need for users to change any code, transferring the workload from machines to Kubernetes.
```

Comment:
Good to see this has been taken into consideration. I understand that there can be specific differences that are framework-specific but how much of this could be common and reused?

Reply:
Thank Terry for reading this consideration, which is the most significant concern in my thinking. Here are some points that I want to discuss with you.

Firstly, let me add one more thing,

**Support 2 partitioning ways**: the operator-wise design part in this proposal is the most ideal implementation for supporting fully distributed partitioning and training. However, distributed partitioning is an experimental new-coming feature, and I think it is a work-around way of DGL recommended, which is not included in DGL library currently. Moreover, the standalone partitioning method is included in DGL library, so I want to support both distributed and standalone in "DGL Operator".

Then, I think if we (all allreduce-way operators) reused MPI Operator as much as possible, MPI Operator may need,

1. Imporve the customerization of the `hostfile` generating formats. DGL Operator's `ipconfig` format: [ip] [dgl server port] [worker num] [host/pod name] [gpu num]
2. Improve the customization of the path of config_volume
3. How to fulfill standalone partitioning without any training workloads' waiting?
   1. Idea 1: At first, use an initContainer to invoke a `Partitioner` Pod. And then take off all training worker Pod. So we may need to find a way of injecting this initContainer.
   2. Idea 2: Use job status to control the workflow of `Partitioner` / `Worker` Pod's starting phase and releasing phase